### PR TITLE
Add find checks for AMGX/AMGXSolver wrapper and ViennaCL.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR adds cmake checks for AMGX, CUDA, AmgXSolver wrapper and ViennaCL.  It also fixes two minor issues for dune-grid and dune-fem checks.

### CODE REVIEW

File: cmake/Modules/FindAmgXSolver.cmake [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    NO_DEFAULT_PATH
+    DOC "Include directory of AMGX")
+
+  find_library(AMGX_LIBRARY
+      NAMES "amgxsh" "amgx"
+      PATHS ${AMGX_ROOT}
+      PATH_SUFFIXES "lib" "lib" "lib/${CMAKE_LIBRARY_ARCHITECTURE}"
+      NO_DEFAULT_PATH)
+
+  find_path(AMGX_WRAPPER_INCLUDE_DIR AmgXSolver.hpp
```
Comment (blattms) [inline cmake/Modules/FindAmgXSolver.cmake]: What is this wrapper? Did not find in my version of AMGX

File: cmake/Modules/FindAmgXSolver.cmake [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+if( NOT PETSC_FOUND )
+  return()
+endif()
+
+set(AMGX_VALID_COMPONENTS
+    C
+    CXX)
+
+SET(CUDA_DIR "$ENV{CUDA_DIR}" CACHE PATH "The path to CUDA.")
+find_package(CUDA)
```
Comment (blattms) [inline cmake/Modules/FindAmgXSolver.cmake]: Shouldn't this be REQUIRED? Or does it work without CUDA.

File: cmake/Modules/FindAmgXSolver.cmake [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# - Try to find AmgX
+# Once done this will define
+#
+#  AMGXSOLVER_FOUND        - system has PETSc
+#  AMGXSOLVER_INCLUDES     - the PETSc include directories
+#  AMGXSOLVER_LIBRARIES    - Link these to use PETSc
+#
+# Redistribution and use is allowed according to the terms of the BSD license.
+# For details see the accompanying COPYING-CMAKE-SCRIPTS file.
```
Comment (blattms) [inline cmake/Modules/FindAmgXSolver.cmake]: Is this copied from somewhere and hence the BSD-Licence? COPYING-CMAKE-SCRIPTS  is not in this PR and probably not applies to other files. Comment seems wrong and needs an update.

File: cmake/Modules/FindAmgXSolver.cmake [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+set(AMGX_VALID_COMPONENTS
+    C
+    CXX)
+
+SET(CUDA_DIR "$ENV{CUDA_DIR}" CACHE PATH "The path to CUDA.")
+find_package(CUDA)
+
+if(${CUDA_FOUND})
+  SET(CUDA_LIBRARIES ${CUDA_LIBRARIES} ${CUDA_CUBLAS_LIBRARIES} ${CUDA_cusparse_LIBRARY} ${CUDA_cusolver_LIBRARY})
+  #MESSAGE("-- Finding CUDA - Success")
```
Comment (blattms) [inline cmake/Modules/FindAmgXSolver.cmake]: When used AMGX in needed additional libraries libdl, librt? Maybe this changed and I should recheck...

File: cmake/Modules/FindViennaCL.cmake [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# - Try to find ViennaCL
+# Once done this will define
+#
+#  VIENNACL_FOUND        - system has ViennaCL
+#  VIENNACL_INCLUDES     - the ViennaCL include directories
+#  OPenCL Variables in addition
+#
+# Redistribution and use is allowed according to the terms of the BSD license.
+# For details see the accompanying COPYING-CMAKE-SCRIPTS file.
```
Comment (blattms) [inline cmake/Modules/FindViennaCL.cmake]: Same holds as for the other file: Copied? Then comment on it. Missing COPYING-CMAKE-SCRIPTS

File: cmake/Modules/FindAmgXSolver.cmake [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#  AMGXSOLVER_LIBRARIES    - Link these to use PETSc
+#
+# Redistribution and use is allowed according to the terms of the BSD license.
+# For details see the accompanying COPYING-CMAKE-SCRIPTS file.
+#
+
+#cmake_policy(VERSION 3.3)
+
+if( NOT PETSC_FOUND )
+  return()
```
Comment (akva2) [inline cmake/Modules/FindAmgXSolver.cmake]: please inform user 'message(STATUS "AMGx disabled, requires PETSc")' or something).

File: cmake/Modules/FindAmgXSolver.cmake [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+endif()
+
+set(AMGX_VALID_COMPONENTS
+    C
+    CXX)
+
+SET(CUDA_DIR "$ENV{CUDA_DIR}" CACHE PATH "The path to CUDA.")
+find_package(CUDA)
+
+if(${CUDA_FOUND})
```
Comment (akva2) [inline cmake/Modules/FindAmgXSolver.cmake]: please stick to modern idioms (do not deref variable).

File: cmake/Modules/FindAmgXSolver.cmake [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  SET(CUDA_LIBRARIES ${CUDA_LIBRARIES} ${CUDA_CUBLAS_LIBRARIES} ${CUDA_cusparse_LIBRARY} ${CUDA_cusolver_LIBRARY})
+  #MESSAGE("-- Finding CUDA - Success")
+  message("Cuda = ${CUDA_LIBRARY_DIRS} ${CUDA_LIBRARIES}")
+
+  # search AMGX header
+  find_path(AMGX_INCLUDE_DIR amgx_c.h amgx_capi.h
+    PATHS ${AMGX_ROOT} ${AMGX_DIR}
+    PATH_SUFFIXES include include/amgx
+    NO_DEFAULT_PATH
```
Comment (akva2) [inline cmake/Modules/FindAmgXSolver.cmake]: hmm? why disable default paths?

File: cmake/Modules/FindAmgXSolver.cmake [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  find_path(AMGX_WRAPPER_INCLUDE_DIR AmgXSolver.hpp
+    PATHS ${AMGX_WRAPPER_ROOT} ${AMGX_WRAPPER_DIR}
+    PATH_SUFFIXES include
+    NO_DEFAULT_PATH
+    DOC "Include directory of AMGX_WRAPPER")
+
+  find_library(AMGX_WRAPPER_LIBRARY
+      NAMES "AmgXWrapper"
+      PATHS ${AMGX_WRAPPER_ROOT}
+      PATH_SUFFIXES "lib" "lib" "lib/${CMAKE_LIBRARY_ARCHITECTURE}"
```
Comment (akva2) [inline cmake/Modules/FindAmgXSolver.cmake]: stick to modern idioms, ie no ROOT, no specifying suffixes. users can use AmgXSolver_DIR

File: cmake/Modules/FindViennaCL.cmake [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#
+
+find_package(OpenCL)
+
+# search VIENNACL header
+find_path(VIENNACL_INCLUDE_DIR version.hpp
+  PATHS ${VIENNACL_ROOT} ${VIENNACL_DIR}
+  PATH_SUFFIXES include include/viennacl
+  NO_DEFAULT_PATH
```
Comment (akva2) [inline cmake/Modules/FindViennaCL.cmake]: same question, why disable default paths?

**Comment by joakim-hove** [comment]
What is the status here?

**Pull request closed without merging**